Add Wyrm and Wyrmlet (amphiptere) to Core (from SoF)

Jonathan-Kelly opened the pull request:
Their statistics are (roughly) aimed at complementing the other small but fast flying creatures in core - bats, falcons, dragonflies. Whereas those other small flyers exclusively use melee attacks, this wyrm line focuses instead on its ranged attack (fire breath). But it is meant to be otherwise comparable to them in general power (and thus less power than the special variants in SoF).

The appearance of SoF's Cave Wyrm and Cave Wymlet are also color graded from green to blue to help distinguish them from these new core variants and to perhaps give them a more subterranean appearance. Other color options include yellow (as with the mythical [amphiptere](https://en.wikipedia.org/wiki/Amphiptere)) or a chalky cave albinism look. Also this experimentally gives the Cave Wyrm and Cave Wyrmlet frost breath to further distinguish them from the other variants (but this is just an idea).

Jonathan-Kelly committed:
SoF: albino Cave Wyrm and Wyrmlet recolor

Jonathan-Kelly committed:
Add core Wyrm and Wyrmlet units

Jonathan-Kelly committed:
WoF: cut redundant Wyrm unit

sevu commented:
Cave Wyrm could be a [variation] as part of the Wyrm unit_type in core.

Jonathan-Kelly commented:
It is probably worth considering at some point (perhaps after the stats of the base unit are more stable). There are some pros and cons to using inheritance for game content.

Hejnewar commented:
Honestly I prefer core units to just be separate and not variations of one another.

doofus-01 commented:
Moving a generic wyrmlet thing to core makes sense, since it is being used in multiple campaigns/scenarios.  

I'm not crazy about the icy cave-type idea, I'd rather they just be a little stronger, better with mushrooms, or something similar.  The ice-type could be saved for some place where it makes special thematic sense, and it could pull assets from core.

If we're going to have so many wyrms & wyrmlets, it would be better if we didn't recycle images.  I don't support the "variations" idea either.

doofus-01 reviewed (changes requested):
Requesting these changes:
1.  Some thematic redesign to the cave wyrm
2.  Firmer outline of what different types are planned and where they go, so we can avoid just recoloring the same images several times

Jonathan-Kelly commented:
> I'm not crazy about the icy cave-type idea,

No problem, removed it.

> The chalky cave albinism isn't a bad idea.

Ran into a small snag with this. I cannot find a way to isolate the drop shadow from the recoloring effect, which prevents much brightening of the sprite without the drop shadow turning grey or pink. The shadow is the only part of the sprite which is semi-transparent, so I tried using a number of `~CHAN` formulae to only affect the fully opaque pixels but it always seems to break the image one way or another.

I was able to color shift the sprite a lighter ashen grey color without affecting the drop shadow too much; what do you think of it? I could also use a the greyscale function to remove the remaining hints of color completely.

> Firmer outline of what different types are planned and where they go,

Only this one type is planned.

CelticMinstrel commented:
> what do you think of it?

I think we need a screenshot to judge…?

> Only this one type is planned.

Though that's entirely fine, it _would_ be cool to build up a whole family of these creatures, I suppose…

CelticMinstrel commented:
> Ran into a small snag with this. I cannot find a way to isolate the drop shadow from the recoloring effect, which prevents much brightening of the sprite without the drop shadow turning grey or pink. The shadow is the only part of the sprite which is semi-transparent, so I tried using a number of `~CHAN` formulae to only affect the fully opaque pixels but it always seems to break the image one way or another.

Another option might be to bake the colour shift using Wesnoth's `--render-image` command-line option, then manually fix the drop shadow.

Jonathan-Kelly commented:
> I think we need a screenshot to judge…?

This is the color shifted version:
![CS_sprites](https://github.com/wesnoth/wesnoth/assets/80054769/5e6415ca-e406-4653-ad5e-327ecae01af0)

The color shifted portrait:
![CS_portrait](https://github.com/wesnoth/wesnoth/assets/80054769/b85f103b-83b3-4f2f-828f-1e0d2067c994)

And this is the grey scale version:
![GS_sprites](https://github.com/wesnoth/wesnoth/assets/80054769/8b4af2c9-a7be-4942-a85a-d7e636fd0cf7)

And the grey scale portrait:
![GS_portrait](https://github.com/wesnoth/wesnoth/assets/80054769/45cbe253-fe9a-4b12-922a-a7a0d74a4196)

For context, this is the Red Wyrm(let) portrait which appears beside the Cave Wyrm(let) in SoF (and is unchanged by this PR):
![red-wyrm_portrait](https://github.com/wesnoth/wesnoth/assets/80054769/4a28130c-a92e-4275-bf0d-ecee42660b15)

CelticMinstrel commented:
The sprite looks alright without greyscale and slightly worse with greyscale, but the portrait looks pretty awful in both versions. The greyscale portrait is definitely better, but it's too dark to be an albino.

I wonder how it would look if you first brightened the portrait, then converted to greyscale? You could also experiment with the sepia filter…

How was the red portrait created? From what I can tell, the base portrait is green, and I'd expect colour shift to make it look bad just like it does for this one, though I could be wrong…

doofus-01 commented:
We can separate the shadow from the body, so it doesn't get shifted.  Better would be to have different images, but if recycling images is ultimately what we end up doing, I think `CHAN()` is probably better for this than `CS()`.  The horses are an example (not that they're ideal either, but at least they have different portraits).

doofus-01 commented:
The sprites look pretty good to me, nice job.  A little crustiness on the wings, but that might be an issue with the original image?  (This is a good example of what pixel art is for.)

The portraits aren't as easy to nicely recycle, but if there's only one or two more wyrm images, it should be manageable for 1.18.

CelticMinstrel commented:
If you're going for albino, that portrait somehow manages to do a pretty good job of portraying it, though it's a bit too green on the creature's back.

The sprite on the other hand looks pretty ugly - I think I liked the previous one better (tho there was obviously the drop shadow issue). The main problem is the yellow wings, but I think it's also a bit _too_ white. Like the portrait, I think there should be some hints of pink. That said, I could live with this whiteness as long as the wings are also the same colour.

Jonathan-Kelly commented:
> The sprites look pretty good to me, nice job. A little crustiness on the wings, but that might be an issue with the original image? (This is a good example of what pixel art is for.)

> The main problem is the yellow wings, but I think it's also a bit too white. Like the portrait, I think there should be some hints of pink. That said, I could live with this whiteness as long as the wings are also the same colour.

I worked out a new formula to improve the wings and lowered the whiteness a bit:

![cave_wyrmlets](https://github.com/wesnoth/wesnoth/assets/80054769/6f6ee581-5859-4318-ad29-a715c6a8c1db)

> If you're going for albino, that portrait somehow manages to do a pretty good job of portraying it, though it's a bit too green on the creature's back.

Here it is with the green reduced a bit:

![cave_wyrmlet](https://github.com/wesnoth/wesnoth/assets/80054769/3556865c-cc34-44c2-9ca3-4746441d4c79)

CelticMinstrel commented:
I guess that looks better. There seems to be some weirdness right at the neck when the wings are forward, but it's pretty close, I guess.

Ultimately I think it'd be best to have separate sprites for these (manually recoloured) instead of using colour-shift, but this colour-shift seems like a reasonable placeholder.

Pentarctagon commented:
Are there any remaining issues with this PR aside from the merge conflicts?

doofus-01 commented:
It should be reviewed against updated master first, not just given git maintenance.

Jonathan-Kelly commented:
Should I hit the "Re-request review" button then or would that just be redundant?

knyghtmare commented:
Has loads of merge conflicts that must be resolved first.

babaissarkar commented:
Probably too late by this point, but what we're calling "Wyrm" is actually "Wyvern" (https://en.wikipedia.org/wiki/Wyvern).

sprunk commented:
No, wyverns have exactly 2 legs.

doofus-01 commented:
> Should I hit the "Re-request review" button then or would that just be redundant?

I guess it couldn't hurt, but from my end the biggest risk is that I'll forget, not that the paperwork isn't in order.   I just want to make sure satisfying git isn't the only thing done before this is merged.  Sorry I didn't get to it this weekend, feel free to bump if nothing happens in a couple weeks.

Jonathan-Kelly committed:
Move Wyrm and Wyrmlet art from SoF to core

Jonathan-Kelly committed:
SoF: use new Wyrm and Wyrmlet art directory

Jonathan-Kelly commented:
It turns out that since this pull request was opened the base wyrmlet portrait changed and a completely new one was added for its wyrm advancement. Interestingly, the old color correction algorithm seems to work okay for the new wyrm portrait but a new one was needed to recreate the albino look from the updated wyrmlet one. Results:

![albino](https://github.com/user-attachments/assets/f471307d-48dd-4977-8de5-cba87984867d)
![alibino2](https://github.com/user-attachments/assets/083f3c2e-32da-4158-bc50-b08fed213397)

Jonathan-Kelly commented:
@doofus-01 so what do you think?

doofus-01 commented:
The wyrm looks like it's glowing; it has no dark tones.  The wyrmlet looks better, but it also looks more overexposed than albino.  I'm not educated enough in color science to predict what would make it better, maybe making the RGB channels darker below a certain threshold, then keeping the highlights a very low saturation red and a bit more muted?

CelticMinstrel commented:
In my opinion, the first one looks reasonably albino, but the second one is far too green.

Jonathan-Kelly commented:
Since the wyrmlet portrait is closer to hitting the mark I will just focus on it for now and then try to apply the resulting look to the wyrm portrait afterwards.

> The wyrmlet looks better, but it also looks more overexposed than albino.

Ah, I see what you mean; the color multipliers were turned up too high so the highlights were effectively blown out. Here is what it looks like with them toned down (and the thick top scales used to white balance it before biasing slightly towards pink:

![wyrmlet](https://github.com/user-attachments/assets/60ab13c6-8a4e-45f6-b2d6-f7bb234e56b5)

> maybe making the RGB channels darker below a certain threshold, then keeping the highlights a very low saturation red and a bit more muted?

Hmm, I could try using the threshold you defined for the red wyrmlet to make the thicker scales creamy white while keeping the thin skin of the undersides darker and pinker. That threshold (or any threshold) may not work as well for this since the red wyrmlet's scales are emissive (glowing like lava) but I will experiment to find out.

By the way, your new red wyrmlet portrait looks fantastic. A real head and shoulders improvement over the original approach.

Jonathan-Kelly commented:
@doofus-01 looks like you are right about using the threshold, it makes all the difference. Here's what the cave wyrmlet looks like now (along with the other two wyrmlets for comparison):
![wyrmlet_cave](https://github.com/user-attachments/assets/7535559e-5f14-4d74-9903-d378da7a40d8)
![wyrmlet](https://github.com/user-attachments/assets/9dc48700-5718-406f-b0c2-461ed8bdce40)
![wyrmlet_red](https://github.com/user-attachments/assets/192d1375-de2f-4de0-b901-a832433e7584)
Do you think this hits the mark?

doofus-01 commented:
That does look much better!

Jonathan-Kelly commented:
And here is the cave wyrm portrait (with the other two for comparison):

![wyrm_cave](https://github.com/user-attachments/assets/b96fe23a-9f8d-435c-b47c-49e6391e0246)
![wyrm](https://github.com/user-attachments/assets/ee342a2d-9205-4093-b3b1-de6ffbfbbd52)
![wyrm_red](https://github.com/user-attachments/assets/9c4e7339-0f70-4aa7-be47-b1b16abb7114)

I think both the cave wyrm and cave wyrmlet portraits are now ready for review.
What are your impressions @doofus-01 and @CelticMinstrel ?

doofus-01 reviewed (approved):
These are recolor, but not a lazy color-shift, I think it reflects better on us and at least some players will appreciate it.  Looks good to go!

Jonathan-Kelly commented:
@doofus-01 do you want the sprite images put directly into `data/core/images/units/monsters` or packaged together into their own sub-directory there - `data/core/images/units/monsters/wyrm` ?

Dalas121 reviewed (changes requested):
Generally looks good to me. Fills a nice missing archetype in our monster lineup.

I'd appreciate some hints in the descriptions about where these units are likely to be found (both which biome and which part of the continent, if any). As our list of fauna continues to grow, I think we'll benefit from some baked-in placement guidelines.

CelticMinstrel commented:
Both look reasonably albino now.

Pentarctagon commented:
@doofus-01 a reminder about this PR.

@Jonathan-Kelly can you address Dalas' last comment?

Jonathan-Kelly commented:
@Dalas121 here are proposals for the three unit descriptions:

**Cave Wyrm**

_"Unlike other varieties of wyrm and bat, the Cave Wyrm dwells only deep underground in places forever untouched by sunlight. Despite its grimly pale appearance, the creature is quite lively - darting through narrow caverns with reckless abandon, gnashing wildly at prey and spitting wisps of flame when threatened by an intruder."_

**Red Wyrm**

_"Much rarer than the common wyrm and almost never seen far from its home caldera, the Red Wyrm is an omininous sight to behold, as a flock overhead often foreshadows the eruption of a nearby volcano. Red Wyrms breathe fire with a keen aim and surprising potency for their size, as if they are themselves tiny dragons."_

**Wyrm** (Common)

_"Wyrms are small, winged, fire breathing serpents native to the hottest regions of the far south, as well as colder climes near to volcanoes, wherein the creatures may overwinter."_

Dalas121 commented:
Descriptions look good to me.

doofus-01 commented:
> @doofus-01 do you want the sprite images put directly into data/core/images/units/monsters or packaged together into their own sub-directory there - data/core/images/units/monsters/wyrm ?

Sorry I missed this.  I think sub-directory would be best.

Jonathan-Kelly committed:
Add unique descriptions for wyrm unit lines

Jonathan-Kelly commented:
@doofus-01 looks like a new CI test is giving this PR a metadata error on the color map/palette PNG file:

`Warning: [Minor] Tag 'XMP:Creator' not defined - data/core/images/units/monsters/wyrm/Wyrm_ColorMap.png
no Creator XMP tag in data/core/images/units/monsters/wyrm/Wyrm_ColorMap.png`

Note this file is only getting moved in this PR, not created or modified.

It looks like GIMP can handle writing IPTC metadata in its metadata editor but I am not sure about (embedded?) XMP or EXIF metadata (though it can read them). What tool do you use to embed this metadata in the image files?

doofus-01 commented:
I use exiftool, but GIMP should be able to write the XMP data.  I hadn't bulk updated the existing PNGs because I wanted to be sure the update was correct before touching so many files, but it has dragged on longer than I intended.

I thought I had written a guide on this, but I'm not finding it now, so...  who knows.  If the metadata check from file move is the only thing holding this up, leave the file where it is without the move and open an issue assigned to me.  Sorry for the complication.

Jonathan-Kelly commented:
> I hadn't bulk updated the existing PNGs because I wanted to be sure the update was correct before touching so many files,

Are there any PNGs with embedded metadata in mainline currently? I am looking for something I could use as a syntax example. I looked at a bunch of random core unit sprite frames in GIMP's metadata viewer but did not find any with author or license data attached.

Also, how would you like to be credited and what license do you want your artwork under (maybe CC-BY-SA) ?

doofus-01 commented:
> Are there any PNGs with embedded metadata in mainline currently? I am looking for something I could use as a syntax example.

https://wiki.wesnoth.org/ImageMetadata is the best reference for now, there is an example at the bottom.  XMP will probably be dropped in favor of Exif, but that hasn't happened yet.

> Also, how would you like to be credited and what license do you want your artwork under (maybe CC-BY-SA) ?

"(doofus-01)" and CC BY-SA 4.0 is good.  Thanks

Jonathan-Kelly committed:
metadata test

Jonathan-Kelly commented:
> XMP will probably be dropped in favor of Exif, but that hasn't happened yet.

GIMP is adding a lot of crap metadata to each sprite frame image, is this a problem or not really?

The metadata:

```<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="XMP Core 4.4.0-Exiv2">
 <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
  <rdf:Description rdf:about=""
    xmlns:xmpMM="http://ns.adobe.com/xap/1.0/mm/"
    xmlns:stEvt="http://ns.adobe.com/xap/1.0/sType/ResourceEvent#"
    xmlns:dc="http://purl.org/dc/elements/1.1/"
    xmlns:GIMP="http://www.gimp.org/xmp/"
    xmlns:tiff="http://ns.adobe.com/tiff/1.0/"
    xmlns:xmp="http://ns.adobe.com/xap/1.0/"
   xmpMM:DocumentID="gimp:docid:gimp:08e8b552-47f4-4753-86d3-83181b568f59"
   xmpMM:InstanceID="xmp.iid:ca296861-9024-48cc-96e6-447cfb21d8f2"
   xmpMM:OriginalDocumentID="xmp.did:e946a725-bd6b-46bf-9a8d-a2f8a7d2f109"
   dc:Format="image/png"
   GIMP:API="3.0"
   GIMP:Platform="Windows"
   GIMP:TimeStamp="1744117923816076"
   GIMP:Version="3.0.2"
   tiff:Orientation="1"
   xmp:CreatorTool="GIMP"
   xmp:MetadataDate="2025:04:08T09:10:04-04:00"
   xmp:ModifyDate="2025:04:08T09:10:04-04:00">
   <xmpMM:History>
    <rdf:Seq>
     <rdf:li
      stEvt:action="saved"
      stEvt:changed="/metadata"
      stEvt:instanceID="xmp.iid:ef8c22d3-fe97-466d-9c77-b0a3358ac93a"
      stEvt:softwareAgent="GIMP 3.0.2 (Windows)"
      stEvt:when="2025-04-08T09:09:53-04:00"/>
     <rdf:li
      stEvt:action="saved"
      stEvt:changed="/"
      stEvt:instanceID="xmp.iid:b5885aa2-ef0d-46b3-b23c-7779eb40ae75"
      stEvt:softwareAgent="GIMP 3.0.2 (Windows)"
      stEvt:when="2025-04-08T09:12:03-04:00"/>
    </rdf:Seq>
   </xmpMM:History>
   <dc:creator>
    <rdf:Seq>
     <rdf:li>(doofus-01)</rdf:li>
    </rdf:Seq>
   </dc:creator>
   <dc:rights>
    <rdf:Alt>
     <rdf:li xml:lang="x-default">CC BY-SA 4.0</rdf:li>
    </rdf:Alt>
   </dc:rights>
  </rdf:Description>
 </rdf:RDF>
</x:xmpmeta>```

doofus-01 commented:
It's a problem in that it would blow up the collected images file size, but it's easy to strip out.  As long as the needed tags are there, cleanup can be scripted, and that's something the merger would need to worry about, not the contributor.

Pentarctagon commented:
Any update for this?

Jonathan-Kelly commented:
> Any update for this?

The push to have a wyrm in core was somewhat predicated on using it for a mainline multiplayer scenario which has since been put on the back burner. And now there is `data/internal` which is a better target for a campaign-only unit which appears in more than one campaign. So maybe it should be redirected there instead ?

Also, this PR contains string changes so it cannot be merged during the current freeze.
